lib/gocodegen.ml

```diff
@@ -91,67 +91,104 @@ let ensure_unique_identifiers (global_t : global_t) =
     (participant, new_role) =
   let local_protocol =
     lookup_local_protocol protocol_lookup protocol new_role
   in
-  let env, role_chan =
-    LTypeCodeGenEnv.new_send_role_channel env participant local_protocol
-      new_role protocol
-  in
-  let env, invite_chan =
-    LTypeCodeGenEnv.new_send_invite_channel env participant local_protocol
+  let env, role_chan, invite_chan =
+    LTypeCodeGenEnv.get_or_add_send_invitation_channels env participant
+      local_protocol new_role protocol
   in
+  (* TODO: Fix label channels *)
   (env, (role_chan, invite_chan))
 
 (** Generate the struct field names for the channels over which to accept the
     invitations from the caller role *)
 let gen_accept_invite_chan_names env curr_role caller new_role protocol
     local_protocol =
   if RoleName.equal caller curr_role then
-    let role_chan =
-      LTypeCodeGenEnv.send_self_role_channel env curr_role local_protocol
-    in
-    let invite_chan =
-      LTypeCodeGenEnv.send_self_invite_channel env curr_role local_protocol
-    in
-    (env, role_chan, invite_chan)
+    LTypeCodeGenEnv.get_or_add_send_invitation_channels env curr_role
+      local_protocol new_role protocol
   else
-    let env, role_chan =
-      LTypeCodeGenEnv.new_recv_role_channel env caller local_protocol
-        new_role protocol
-    in
-    let env, invite_chan =
-      LTypeCodeGenEnv.new_recv_invite_channel env caller local_protocol
-    in
-    (env, role_chan, invite_chan)
+    LTypeCodeGenEnv.get_or_add_recv_invitation_channels env caller
+      local_protocol new_role protocol
 
 (** Extract labels from the first interactions in each local type. These
     labels will not necessarily be unique in the case of protocol calls *)
 let extract_choice_labels ltypes =
-  let extract_label = function
+  let rec extract_label = function
     | InviteCreateL (_, _, protocol, _) ->
         LabelName.of_string @@ ProtocolName.user protocol
     | SendL ({label; _}, _, _) -> label
+    | MuL (_, _, ltype) -> extract_label ltype
+    | TVarL _ ->
+        raise
+          (Err.Violation
+             "Currently unfolding of recursion not supported for extracting \
+              choice labels - branches of choices should have an explit \
+              first message")
     | _ ->
-        failwith
-          "First choice interaction should always be a Send or InviteCreate"
+        raise
+          (Err.Violation
+             "Cannot generate code for nested choices - choices should be \
+              flattened or local type should be normalised to ensure this")
   in
   List.map ~f:extract_label ltypes
 
+let extract_choice_label_enums msgs_env ltypes =
+  let rec extract_enum_label = function
+    | AcceptL (_, protocol, roles, _, _, _) ->
+        MessagesEnv.get_invitation_enum msgs_env protocol roles
+    | RecvL ({label; _}, _, _) -> MessagesEnv.get_message_enum msgs_env label
+    | _ ->
+        raise
+          (Err.Violation
```

Comment:
There is a new function in `Err` called `violation` which basically raises a `Violation` (saves typing `raise`),
and another functional called `violationf` which is `violation` with `sprintf` for your convenience.